### PR TITLE
[JENKINS-45256] The reload-configuration CLI command ought to wait until the reload is finished

### PR DESCRIPTION
**Pull request by jglick**
### Proposed changelog entries

* JENKINS-45256 - The `reload-configuration` CLI command now waits until the reload is finished, and returns an error code if the reload failed.

### Desired reviewers

@reviewbybees

### CODE REVIEW

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 1]
```diff
@@ -3986,7 +3986,7 @@ public synchronized HttpResponse doReload() throws IOException {
         LOGGER.log(Level.WARNING, "Reloading Jenkins as requested by {0}", getAuthentication().getName());
 
         // engage "loading ..." UI and then run the actual task in a separate thread
-        servletContext.setAttribute("app", new HudsonIsLoading());
+        WebApp.get(servletContext).setApp(new HudsonIsLoading());
```
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: Not actually necessary but noticed while I was here that this code was failing to use the official accessor API.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 2]
```diff
@@ -4025,7 +4025,7 @@ public void reload() throws IOException, InterruptedException, ReactorException
 
         User.reload();
         queue.load();
-        servletContext.setAttribute("app", this);
+        WebApp.get(servletContext).setApp(this);
```
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: Ditto.

File: core/src/main/java/hudson/cli/ReloadConfigurationCommand.java [hunk 3]
```diff
@@ -43,8 +46,26 @@ public String getShortDescription() {
 
     @Override
     protected int run() throws Exception {
-        Jenkins.getActiveInstance().doReload();
-        return 0;
+        Jenkins j = Jenkins.getInstance();
+        // Or perhaps simpler to inline the thread body of doReload?
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/cli/ReloadConfigurationCommand.java]: the current version works for me

File: core/src/main/java/hudson/cli/ReloadConfigurationCommand.java [hunk 4]
```diff
@@ -43,8 +46,26 @@ public String getShortDescription() {
...
+        j.doReload();
+        Object app;
+        while ((app = WebApp.get(j.servletContext).getApp()) instanceof HudsonIsLoading) {
+            Thread.sleep(100);
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/cli/ReloadConfigurationCommand.java]: Should be a timeout there? I know cases when Jenkins infinitely hangs on reloading. Not sure it is safe to just leave it as is then

**Review by oleg-nenashev — APPROVED** [review]
The change looks good to me, but it definitely smells like a bug. JIRA ticket is required IMHO, maybe there are existing ones

**Comment by jglick** [comment]
> it definitely smells like a bug

I do not think so—@pjanouse just did not think of making the command blocking, like CLI commands normally are.

File: core/src/main/java/hudson/cli/ReloadConfigurationCommand.java (hunk 4 above)
Comment (jglick) [inline core/src/main/java/hudson/cli/ReloadConfigurationCommand.java]: Well, if the command hangs then someone will kill it. I see no need for an internal timeout.

**Comment by oleg-nenashev** [comment]
Created https://issues.jenkins-ci.org/browse/JENKINS-45256

**Pull request merged**